resources/type-system/inference.md

```diff
@@ -1210,6 +1247,674 @@ of steps:
+known as _argument part inference_. _Argument part inference_ is a type
+inference step that is applied to an optional target function type `F`, a
+sequence of expressions `{e_1, e_2, ...}`, a corresponding sequence (of the same
+length) of optional name identifiers `{n_1, n_2, ...}`, a sequence of zero or
+more type arguments `{T_1, T_2, ...}`, and a type schema `K`, known as the
+context. Since the target function type and the names are optional, we will use
+the symbol `∅` to denote a missing target function type or a missing name. _A
+sequence of zero type arguments is typically used when supplying arguments to a
+non-generic function or method, or when type arguments need to be inferred._
+
+The output of argument part inference is a sequence of elaborated expressions
+`{m_1, m_2, ...}` (of the same length as the sequence of input expressions), a
+sequence of zero or more elaborated type arguments `{U_1, U_2, ...}`, and a
+result type known as `R`. _(The sequence of optional names is unchanged by
+argument part inference.)_
+
+To emphasize the relationship between argument part inference and the syntax of
+Dart source code, the inputs and outputs of argument part inference are
+sometimes described using the _&lt;argumentPart&gt;_ syntax, namely `<T_1, T_2,
+...>(n_1: e_1, n_2: e_2, ...)`.
+
+_So, for example, if we say that "argument part inference is invoked on `<int,
+String>(1, x: 2)`", this means that the input of argument part inference is
+applied to the sequence of expressions `{1, 2}`, the sequence of optional names
+`{∅, x}`, and the sequence of type arguments `{int, String}`._
+
+The procedure for argument part inference is as follows:
+
+- Let `{X_1, X_2, ...}` be the list of formal type parameters in `F`, or an
+  empty list if `F` is `∅`.
+
+- Let `{P_1, P_2, ...}` be the list of formal parameter types corresponding to
+  `{e_1, e_2, ...}`, determined as follows:
+
+  - If `F` is `∅`, then let each `P_i` be `_`.
+
+  - Otherwise:
+
+    - For each `n_i` that is `∅`, let `P_i` be the type of the corresponding
+      positional parameter in `F`. If there is no such parameter, it is a
+      compile time error (_too many positional arguments supplied_).
+
+    - For each `n_i` that is not `∅`, let `P_i` be the type of the parameter of
+      `F` named `n_i`. If there is no such parameter, it is a compile time error
+      (_unnexpected named argument supplied_).
+
+- Let `R_F` be the return type of `F`. Or, if `F` is `∅`, then let `R_F` be
+  `dynamic`.
+
+- Produce an initial list of type constraints `C`, and an initial list of type
+  schemas `{U_1, U_2, ...}`, as follows:
+
+  - If `F` is `∅`, then:
+
+    - Let `C` be an empty list of type constraints, and let `{U_1, U_2, ...}` be
+      the same as `{T_1, T_2, ...}`. _This covers the case of dynamic
+      invocation; the user-supplied type arguments are accepted without
+      modification._
+
+  - Otherwise, `F` must be a function type. If the number of formal type
```

Comment:
I wonder whether it would also be helpful to just split out the case where type arguments are explicitly passed into a separate section as well?  There will be some redundancy, but I don't think a lot, and it would make all of this much more readable.  Basically, instead of re-branching on which case you're in many times throughout the description (which is super hard to read), you'd end up with something that has the structure:

if we're in the dynamic case, do {
 downwards inference on arguments using `_`
} else if we're in the explicit type argument case, do {
  compute the contexts for downward inference `K_i`
  downwards inference on arguments using `K_i`
} else if we're in the type inference case do {
  stuff (with some overlap from the above)
}

Reply:
Good point. I’ll attempt this in a follow-up PR as well.